rotkehlchen/chain/evm/decoding/zero_x/decoder.py

```diff
@@ -0,0 +1,208 @@
+        self.native_currency = native_currency
+
+    def _update_send_and_receive_events(
+            self,
+            send_event: 'EvmEvent | None' = None,
+            receive_event: 'EvmEvent | None' = None,
+    ) -> None:
+        if send_event is not None:
+            send_event.counterparty = CPT_ZERO_X
+            send_event.address = self.router_address
+            send_event.event_type = HistoryEventType.TRADE
+            send_event.event_subtype = HistoryEventSubType.SPEND
+            send_event.notes = f'Swap {send_event.balance.amount} {send_event.asset.symbol_or_name()} in 0x'  # noqa: E501
+        if receive_event is not None:
+            receive_event.counterparty = CPT_ZERO_X
+            receive_event.address = self.router_address
+            receive_event.event_type = HistoryEventType.TRADE
+            receive_event.event_subtype = HistoryEventSubType.RECEIVE
+            receive_event.notes = f'Receive {receive_event.balance.amount} {receive_event.asset.symbol_or_name()} as the result of a swap in 0x'  # noqa: E501
+
+    def _decode_swap(
+            self,
+            transaction: 'EvmTransaction',
+            decoded_events: 'list[EvmEvent]',
+            all_logs: 'list[EvmTxReceiptLog]',
+    ) -> 'list[EvmEvent]':
+        """This function is used to decode the swaps done via 0x. It checks if already decoded
+        events interacted with the 0x router, and overwrites them if they did."""
+        send_address_to_events: dict[ChecksumEvmAddress, 'EvmEvent'] = {}
+        receive_address_to_events: dict[ChecksumEvmAddress, 'EvmEvent'] = {}
+        sent_asset = received_asset = None
+        for event in decoded_events:
+            if (
+                event.location_label is None or event.location_label != transaction.from_address or
+                not self.base.is_tracked(string_to_evm_address(event.location_label))
+            ):
+                continue
+
+            if ((
+                event.event_type == HistoryEventType.TRADE and  # for token to token
+                event.event_subtype == HistoryEventSubType.SPEND
+            ) or (
+                event.event_type == HistoryEventType.SPEND and  # for eth to token
+                event.event_subtype == HistoryEventSubType.NONE
+            )) and event.address is not None:
+                sent_asset = event.asset
+                send_address_to_events[event.address] = event
+
+            if ((
+                event.event_type == HistoryEventType.TRADE and  # for token to token
+                event.event_subtype == HistoryEventSubType.RECEIVE
+            ) or (
+                event.event_type == HistoryEventType.RECEIVE and  # for token to eth
+                event.event_subtype == HistoryEventSubType.NONE
+            )) and event.address is not None:
+                received_asset = event.asset
+                receive_address_to_events[event.address] = event
+
+        if sent_asset is None or received_asset is None:
```

Comment:
It's the second case. One possibility is if 0x swapped it in some LP other than uniswap. Then it's a case that we don't support yet. I think we should not log it because this part can be reached if the transaction is entirely not on 0x.